### PR TITLE
Allow fractional values of `'warmup_steps'` option

### PR DESCRIPTION
**Pull request by jukofyork**
Fractional values less than 1 are converted into "fraction of epoch" worth of steps.

Useful to keep the proportion of `'warmup_steps'` constant while altering things like `gradient_accumulation_steps`, etc.

### CODE REVIEW

File: train.py [hunk 1]
```diff
@@ -548,7 +553,7 @@ def get_optimizer(model_parameters):
 
     load_optimizer_states = config.get('load_optimizer_states', True)
     # if resuming and not loading optimizer states, we can't use warmup or the LR never changes from the initial value (still don't know why)
-    if 'warmup_steps' in config and config['warmup_steps'] > 0 and load_optimizer_states:
+    if warmup_steps > 0 and load_optimizer_states:
         warmup_steps = config['warmup_steps']
```
Comment (kallewoof) [inline train.py:557]: I think you want to delete the line here, as it overrides the new value with the one with fractions.

Comment (jukofyork) [reply train.py]: Nice catch!

It was a copy and paste fail from [here](https://github.com/jukofyork/qlora-pipe/blob/decoupled-orthogonal-regularisation/train.py) file trying to use GitHub online editor :frowning:

**Review by kallewoof — APPROVED** [review]
LGTM

**Comment by jukofyork** [comment]
Probably not something to add to `qlora-pipe` but an interesting titbit for anyone interested:

If you set $\beta_1 = 0$ for Adam you get something very similar to [RMSprop](https://pytorch.org/docs/stable/generated/torch.optim.RMSprop.html):

- It still has the "bias correction" of Adam which generally doesn't seem to be used for RMSProp AFAIK.
- It is the "non-centred" version (see the `centered` option for the pytorch docs or section 4.2 of Alex Graves's famous [Generating Sequences With Recurrent Neural Networks](https://arxiv.org/abs/1308.0850) paper).

So the question then is: "what is the optimal warm-up schedule for this?"

Well it turns out if you make lots of (highly unrealistic!) assumptions, the [coefficient of variation](https://en.wikipedia.org/wiki/Coefficient_of_variation) of the bias-corrected second moment estimate $\hat{v}_t$ is:

$$
\text{CV}[ \hat{v}_t ] = \sqrt{ \frac{2 (1 + \beta_2^t)}{1 - \beta_2^t} }
$$

ie: this expression tells us how the relative error (uncertainty) in our estimate of the variance changes over time \(t\) for a given $\beta_2$, due to the way the [exponential moving average ](https://en.wikipedia.org/wiki/Exponential_smoothing) has less samples going into its "memory" at the start of training (**IF** the highly unrealistic! assumptions about stationary, normally distributed errors and so on were true...). 

In the limit:

$$
\text{CV}[ \hat{v}_{t \to \infty} ] = \sqrt{2}
$$

So if we take the reciprocal of this, and then divide by $\sqrt{2}$ we can use this as a learning rate schedule:

image

You can plug other values for $\beta_2$ in [here](https://www.wolframalpha.com/input?i=plot+Sqrt%5B%281+-+%5Cbeta%5Et%29%2F%281+%2B+%5Cbeta%5Et%29%5D+where+%5Cbeta%3D0.99+and+t+between+1+and+500).

and you can use `torch.optim.lr_scheduler.LambdaLR` to implement it very easily:

```python
    elif config['lr_scheduler'] == 'beta2_coupled_lr':
        beta = config['optimizer']['beta2']
        
        def make_rms_ratio_fn(beta):
            def rms_ratio_fn(step):
                return torch.sqrt(torch.tensor((1 - beta**step)/(1 + beta**step))).item()
            return rms_ratio_fn
        
        lr_scheduler = torch.optim.lr_scheduler.LambdaLR(
            optimizer,
            lr_lambda=make_rms_ratio_fn(beta)
        )
```

Anyway, not sure if anybody is interested, and you can probably do very similar by just using the current `torch.optim.lr_scheduler.LinearLR` scheduler via `warmup_steps` (ie: try to find the value of `warmup_steps` that approximates the early part; say `warmup_steps = 200` for $\beta_2 = 0.99$, etc).
